test/sys/test_socket.rs

```diff
@@ -568,12 +569,142 @@ pub fn test_recv_ipv4pktinfo() {
+                    }
+                    _ => continue,
+                }
+            }
+        }
+        None
+    }
+
+    let lo_ifaddr = loopback_v4addr();
+    let (lo_name, lo) = match lo_ifaddr {
+        Some(ifaddr) => (ifaddr.interface_name,
+                         ifaddr.address.expect("Expect IPv4 address on interface")),
+        None => return,
+    };
+    let receive = socket(
+        AddressFamily::Inet,
+        SockType::Datagram,
+        SockFlag::empty(),
+        None,
+    ).expect("receive socket failed");
+    bind(receive, &lo).expect("bind failed");
+    let sa = getsockname(receive).expect("getsockname failed");
+    setsockopt(receive, Ipv4RecvIf, &true).expect("setsockopt IP_RECVIF failed");
+    setsockopt(receive, Ipv4RecvDstAddr, &true).expect("setsockopt IP_RECVDSTADDR failed");
+
+    let thread = thread::spawn(move || {
+        let mut buf = [0u8; 8];
+        let iovec = [IoVec::from_mut_slice(&mut buf)];
+        let mut space = CmsgSpace::<(libc::sockaddr_dl, CmsgSpace<libc::in_addr>)>::new();
+        let msg = recvmsg(
+            receive,
+            &iovec,
+            Some(&mut space),
+            MsgFlags::empty(),
+        ).expect("recvmsg failed");
+        assert!(
+            !msg.flags
+                .intersects(MsgFlags::MSG_TRUNC | MsgFlags::MSG_CTRUNC)
+        );
+
+        for cmsg in msg.cmsgs() {
+            match cmsg {
+                ControlMessage::Ipv4RecvIf(dl) => {
+                    let i = if_nametoindex(lo_name.as_bytes()).expect("if_nametoindex");
+                    assert_eq!(
+                        dl.sdl_index as libc::c_uint,
+                        i,
+                        "unexpected ifindex (expected {}, got {})",
+                        i,
+                        dl.sdl_index
+                    );
+                },
+                ControlMessage::Ipv4RecvDstAddr(addr) => {
+                    if let SockAddr::Inet(InetAddr::V4(a)) = lo {
+                        assert_eq!(a.sin_addr.s_addr,
+                                   addr.s_addr,
+                                   "unexpected destination address (expected {}, got {})",
+                                   a.sin_addr.s_addr,
+                                   addr.s_addr);
+                    } else {
+                        panic!("unexpected Sockaddr");
+                    }
+                },
+                _ => panic!("unexpected additional control msg"),
+            }
+        }
+        assert_eq!(
+            iovec[0].as_slice(),
+            [1u8, 2, 3, 4, 5, 6, 7, 8]
+        );
+    });
+
+    let slice = [1u8, 2, 3, 4, 5, 6, 7, 8];
+    let iov = [IoVec::from_slice(&slice)];
+
+    let send = socket(
+        AddressFamily::Inet,
+        SockType::Datagram,
+        SockFlag::empty(),
+        None,
+    ).expect("send socket failed");
+    sendmsg(send, &iov, &[], MsgFlags::empty(), Some(&sa)).expect("sendmsg failed");
+
+    thread.join().unwrap();
```

Comment:
I don't think you need a separate thread.  sendmsg(2) should return before recvmsg(2) gets called.